force-app/main/default/classes/ContactController.cls

```diff
@@ -1,170 +1,132 @@
+        for (Contact c : source) {
+            result.add(c.AccountId);
         }
+        return new List<Id>(result);
+    }
+
+    private static Map<Id, List<Contact>> getColleagues(List<Contact> source, List<Id> accountsId){
+        List<Contact> temp = [SELECT Id, AccountId FROM Contact c WHERE c.AccountId IN :accountsId AND Id NOT IN :source];
+        Map<Id, List<Contact>> result = new Map<Id, List<Contact>>();
+        for (Contact c : temp) {
+            if (result.keySet().contains(c.AccountId)) {
+                List<Contact> tmp = result.get(c.AccountId);
+                tmp.add(c);
+            }else {
+                result.put(c.AccountId, new List<Contact>{c});
+            }
+        }        
         return result;
-    } 
+    }
 
-    private static Boolean isContactHasCases(List<Case> source, Id contactId){
-        Boolean result = true;
-        List<Case> contactCases = CaseController.findCasesByContactId(source, contactId, allStatuses);
-        if (contactCases == null) {
-            result = false;
+    private static List<Id> getAllContactsId(List<Contact> delContacts, Map<Id, List<Contact>> accountWithContacts){
+        List<Id> result = new List<Id>();
+        List<Contact> tempContacts = new List<Contact>();
+        for(List<Contact> temp : accountWithContacts.values()){
+            tempContacts.addAll(temp);
+        }
+        tempContacts.addAll(delContacts);
+        for(Contact c : tempContacts){
+            result.add(c.Id);
         }
         return result;
-    }   
+    }
 
-    private static Contact findContactWithMinCases (Id contactId){        
-        List<Contact> source = findContactColleagues(contactId);        
-        if (source.size() == 0) {
-            return null;
-        }else {   
-            Contact result;  
-            Map<Id, Contact> contactsWithCountCases = new Map<Id, Contact>(source);
-            List<Contact> resList = [SELECT Id, (SELECT Id FROM Cases) FROM Contact WHERE Id IN :contactsWithCountCases.keySet()];
-            Map<Contact, Integer> resMap = new Map<Contact, Integer>();
-            for(Contact c : resList){
-                resMap.put(c, c.Cases.size());
-            }           
-            List<Contact> contactList = new List<Contact>(resMap.keySet());
-            List<Integer> caseQuantity = resMap.values();
-            Integer count = 2147483646;
-            Integer tmp;
-            for(Integer i = 0; i < caseQuantity.size(); i++){
-                if (caseQuantity.get(i) < count) {
-                    count = caseQuantity.get(i);
-                    tmp = i;
-                }
+    private static Map<Contact, List<Case>> getAllCases(List<Id> allContactsId){
+        List<Contact> allContacts = [SELECT Id, AccountId, (SELECT Id, Status, ContactId FROM Cases) FROM Contact WHERE Id IN :allContactsId];
+        Map<Contact, List<Case>> result = new Map<Contact, List<Case>>();
+        for(Contact c : allContacts){
```

Comment:
you can directly put the SOQL from 91 line instead of "allContacts" here, no need to keep one more list in heap

Reply:
Fixed